tests/admin_inlines/tests.py

```diff
@@ -228,6 +229,113 @@ def test_inline_hidden_field_no_column(self):
             response.rendered_content,
         )
 
+    def test_tabular_inline_hidden_field_with_view_only_permissions(self):
+        """#31867 -- Make sure content of hidden field is not visible in tabular inline when user has view-only
+        permission"""
+        parent = Parent1.objects.create(name='a')
+        Child1.objects.create(name='b', position='0', parent=parent)
+        Child1.objects.create(name='c', position='1', parent=parent)
+        user = User.objects.create_user(username="user", password="pwd", is_staff=True)
+        parent1_ct = ContentType.objects.get_for_model(Parent1)
+        child1_ct = ContentType.objects.get_for_model(Child1)
+        permission = Permission.objects.get(codename='view_parent1', content_type=parent1_ct)
+        user.user_permissions.add(permission)
+        permission = Permission.objects.get(codename='view_child1', content_type=child1_ct)
+        user.user_permissions.add(permission)
+        self.client.login(username="user", password="pwd")
+        response = self.client.get(reverse('admin:admin_inlines_parent1_change', args=(parent.pk,)))
+        self.assertInHTML('<th class="column-position hidden">Position</th>', response.rendered_content)
+        self.assertInHTML('<td class="field-position hidden"><p>0</p></td>', response.rendered_content)
+        self.assertInHTML('<td class="field-position hidden"><p>1</p></td>', response.rendered_content)
+
+    def test_stacked_inline_hidden_field_with_view_only_permissions(self):
+        """#31867 -- Make sure content of hidden field is not visible in stacked inline when user has view-only
+        permission"""
+        parent = Parent2.objects.create(name='a')
+        Child2.objects.create(name='b', position='0', parent=parent)
+        Child2.objects.create(name='c', position='1', parent=parent)
+        user = User.objects.create_user(username="user", password="pwd", is_staff=True)
+        parent2_ct = ContentType.objects.get_for_model(Parent2)
+        child2_ct = ContentType.objects.get_for_model(Child2)
+        permission = Permission.objects.get(codename='view_parent2', content_type=parent2_ct)
+        user.user_permissions.add(permission)
+        permission = Permission.objects.get(codename='view_child2', content_type=child2_ct)
+        user.user_permissions.add(permission)
+        self.client.login(username="user", password="pwd")
+        response = self.client.get(reverse('admin:admin_inlines_parent2_change', args=(parent.pk,)))
+        # Following test ensure that the whole line containing name + position fields is not hidden
+        self.assertContains(response, '<div class="form-row field-name field-position">')
+        # Following test ensure that the div containing the position field is hidden
```

Comment:
Django style is to avoid "ensure that" and similar boilerplate language; just say "The whole line containing ... is not..."